facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+        // We may not yet be at the last chunk, and hence may not know the total
+        // length of eventual complete object, but this variable the total
+        // length of all the bytes fed into write(), even if we haven't uploaded
+        // them all yet.
+        let object_length_thus_far = self.object_upload_position + chunk_len;
+
+        // When this is the last piece being uploaded, the Content-Range header
+        // should include the total object size, but otherwise should have * to
+        // indicate to GCS that there is an unknown further amount to come.
+        // https://cloud.google.com/storage/docs/streaming#streaming_uploads
+        let content_range_header_total_length_field = if last_chunk {
+            format!("{}", object_length_thus_far)
+        } else {
+            "*".to_owned()
+        };
+        let content_range = format!(
+            "bytes {}-{}/{}",
+            self.object_upload_position,
+            object_length_thus_far - 1,
+            content_range_header_total_length_field
+        );
+
+        // ureq::Request::send takes possession of the argument so we must clone
+        // here, in case the entire chunk doesn't get uploaded successfully.
+        let body = Cursor::new(chunk.clone());
+
+        let http_response = ureq::put(&self.upload_session_uri)
+            .set("Content-Length", &format!("{}", chunk_len))
+            .set("Content-Range", &content_range)
+            // By default, ureq will wait forever to connect or read
+            .timeout_connect(10_000) // ten seconds
+            .timeout_read(10_000) // ten seconds
+            .send(body);
+
+        // On success we expect HTTP 308 Resume Incomplete and a Range: header,
+        // unless this is the last part and the server accepts the entire
+        // provided Content-Range, in which case it's HTTP 200, or 201 (?).
+        // https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+        match http_response.status() {
+            200 | 201 => {
+                if !last_chunk {
+                    return Err(anyhow!(
+                        "received HTTP 200 or 201 response with chunks remaining"
+                    ));
+                }
+            }
+            308 => {
+                if !http_response.has("Range") {
+                    return Err(anyhow!(
+                        "No range header in response from GCS: {:?}",
+                        http_response.into_string()
+                    ));
+                }
+                let range_header = http_response.header("Range").unwrap();
+                // The range header is like "bytes=111-222", and represents a
+                // range in the overall object, not the current chunk.
+                let range = range_header
+                    .strip_prefix("bytes=")
+                    .context(format!("unexpected Range header {}", range_header))?;
+                let mut values = range.splitn(2, '-');
```

Comment:
I think it would simplify the following code to collect into a Vec:

```
let values: Vec<_> = range.splitn(2, '-').collect();
if values.len() != 2 {
 // return error
}
let start: usize = values[0].parse().context("unexpected ...
```